### PR TITLE
Implement RPC for query by raw hash

### LINKED ISSUE #235: Ensure we can query the log by merkle leaf and value hash
Pretty sure that both of these aren't implemented atm as that work predated the split.

### PR DESCRIPTION
**Pull request by Martin2112**
Implement get leaves by leaf value hash in log RPC server and corresponding mock updates. Shares most of the code with the query by Merkle hash API. Closes #235.

### CODE REVIEW

File: server/log_rpc_server_test.go [hunk 1]
```diff
@@ -526,6 +526,118 @@ func TestGetLeavesByHash(t *testing.T) {
...
+	defer ctrl.Finish()
+
+	mockStorage := storage.NewMockLogStorage(ctrl)
+	mockTx := storage.NewMockLogTX(ctrl)
+
+	mockStorage.EXPECT().Begin().Return(mockTx, errors.New("TX"))
+
+	server := NewTrillianLogRPCServer(mockStorageProviderfunc(mockStorage), fakeTimeSource)
+
+	if _, err := server.GetLeavesByLeafValueHash(context.Background(), &getByHashRequest1); err == nil || !strings.Contains(err.Error(), "TX") {
```
Comment (daviddrysdale) [inline server/log_rpc_server_test.go]: Prob'ly worth splitting out the two distinct failure cases here

File: server/log_rpc_server_test.go [hunk 2]
```diff
@@ -526,6 +526,118 @@ func TestGetLeavesByHash(t *testing.T) {
...
+
+	test := newParameterizedTest(ctrl, "GetLeavesByLeafValueHash",
+		func(t *storage.MockLogTX) {
+			t.EXPECT().GetLeavesByLeafValueHash([][]byte{[]byte("test"), []byte("data")}, false).Return([]trillian.LogLeaf{}, errors.New("STORAGE"))
+		},
+		func(s *TrillianLogRPCServer) error {
+			_, err := s.GetLeavesByLeafValueHash(context.Background(), &getByHashRequest1)
+			return err
+		})
+
```
Comment (daviddrysdale) [inline server/log_rpc_server_test.go]: (pre-existing) I have problems understanding what's going on with these tests -- there's quite a lot of action-at-a-distance and mixtures of constants and explicit values.  Maybe discuss offline.

File: server/log_rpc_server.go [hunk 3]
```diff
@@ -587,3 +576,31 @@ func fetchNodesAndBuildProof(tx storage.LogTX, treeRevision, leafIndex int64, pr
...
+// getLeavesByHashInternal does the work of fetching leaves by either their raw data or merkle
+// tree hash depending on the supplied fetch function
+func (t *TrillianLogRPCServer) getLeavesByHashInternal(ctx context.Context, desc string, req *trillian.GetLeavesByHashRequest, fetchFunc func(storage.LogTX, [][]byte, bool) ([]trillian.LogLeaf, error)) (*trillian.GetLeavesByHashResponse, error) {
+	ctx = util.NewLogContext(ctx, req.LogId)
+	if len(req.LeafHash) == 0 || !validateLeafHashes(req.LeafHash) {
+		return &trillian.GetLeavesByHashResponse{Status: buildStatusWithDesc(trillian.TrillianApiStatusCode_ERROR, fmt.Sprintf("%s: Must supply at least one hash and none must be empty", desc))}, nil
+	}
+
+	tx, err := t.prepareStorageTx(req.LogId)
+
```
Comment (Martin2112) [inline server/log_rpc_server.go]: Done. But I think it hides the important thing (assigning to tx) compared to the error handling.

File: server/log_rpc_server_test.go [hunk 4]
```diff
@@ -526,6 +526,118 @@ func TestGetLeavesByHash(t *testing.T) {
...
+	mockStorage := storage.NewMockLogStorage(ctrl)
+
+	server := NewTrillianLogRPCServer(mockStorageProviderfunc(mockStorage), fakeTimeSource)
+
+	// This request includes an empty hash, which isn't allowed
+	resp, err := server.GetLeavesByLeafValueHash(context.Background(), &getByHashRequestBadHash)
+
+	// Should have succeeded at RPC level
+	if err != nil {
+		t.Fatalf("GetLeavesByLeafValueHash() = %v", err)
```
Comment (Martin2112) [inline server/log_rpc_server_test.go]: Done.

File: server/log_rpc_server_test.go [hunk 5]
```diff
@@ -526,6 +526,118 @@ func TestGetLeavesByHash(t *testing.T) {
...
+	}
+
+	// And failed at app level
+	if got, want := resp.Status.StatusCode, trillian.TrillianApiStatusCode_ERROR; got != want {
+		t.Fatalf("Bad status got: %v, want: %v", got, want)
```
Comment (Martin2112) [inline server/log_rpc_server_test.go]: Done.

File: server/log_rpc_server_test.go (hunk 2 above)
Comment (Martin2112) [inline server/log_rpc_server_test.go]: It seemed to make sense at the time. There's a matrix of RPC operations and possible failures of the storage mock. The parameterizedTest wraps up the setup and RPC call and ensures that the API calls we expect are made on storage.

File: server/log_rpc_server_test.go (hunk 4 above)
Comment (Martin2112) [inline server/log_rpc_server_test.go]: Done.

File: server/log_rpc_server_test.go (hunk 1 above)
Comment (Martin2112) [inline server/log_rpc_server_test.go]: Done.

**Comment by Martin2112** [comment]
Sorry, some stuff from another branch managed to get mixed in. Let me sort it out.

**Comment by Martin2112** [comment]
OK, the rogue commit has been dropped. It builds again.

File: server/log_rpc_server_test.go [hunk 6]
```diff
@@ -539,12 +539,12 @@ func TestGetLeavesByLeafValueHashInvalidHash(t *testing.T) {
 
 	// Should have succeeded at RPC level
 	if err != nil {
-		t.Fatalf("GetLeavesByLeafValueHash() = %v", err)
+		t.Fatalf("GetLeavesByLeafValueHash(hash empty) = %v, want nil", err)
```
Comment (daviddrysdale) [inline server/log_rpc_server_test.go]: This is still confusing (here and below) because the error messages don't distinguish between the two return values.

Comment (Martin2112) [inline server/log_rpc_server_test.go]: Not sure what you mean. In this case we want no RPC level error but an app status error. In the case below we expect a TX specific error but the message was wrong after splitting the cases. I've tried again but I don't know if I've addressed the point.

**Comment by Martin2112** [comment]
Github isn't letting me comment on the bytesToHashes entry but you're right. It's been rendered pointless and has been removed.

File: server/log_rpc_server_test.go (hunk 6 above)
Comment (daviddrysdale) [inline server/log_rpc_server_test.go]: Ah, I wasn't clear enough, let me try again.  `GetLeavesByLeafValueHash()` has two return values, a response and an error code.  In some of the tests you're checking one, in other tests you're checking the other, and it was hard to tell which is which from the output alone.

This was why I suggested `"GetLeavesByLeafValueHash(badHash) = %s, %v; want Api_ERROR, nil"` as an error format string -- it shows both returns.

Comment (Martin2112) [inline server/log_rpc_server_test.go]: OK. I thought they'd be distinct because the API level status is an integer not an error but it sounds like I need to do more.

**Comment by Martin2112** [comment]
Not sure what's left to do on this one but I'd like to move it on.

**Comment by daviddrysdale** [comment]
The only minor thing left was just to make the `t.Fatalf` messages be clear about whether it's the first return value or the second error return value that has the wrong result.

**Comment by Martin2112** [comment]
OK I will address that.

File: server/log_rpc_server_test.go [hunk 7]
```diff
@@ -539,12 +539,12 @@ func TestGetLeavesByLeafValueHashInvalidHash(t *testing.T) {
 
 	// Should have succeeded at RPC level
 	if err != nil {
-		t.Fatalf("GetLeavesByLeafValueHash(hash empty) = %v, want: nil", err)
+		t.Fatalf("GetLeavesByLeafValueHash(hash empty) = %v, %v, want: nil, TrillianApiStatusCode_ERROR", err, resp.Status)
```
Comment (daviddrysdale) [inline server/log_rpc_server_test.go]: Aren't both of these the wrong way round?  I.e. shouldn't it read: 
```go
t.Fatalf("GetLeavesByLeafValueHash(hash empty) = %v, %v, want: TrillianApiStatusCode_ERROR, nil",  resp.Status, err)
```

File: server/log_rpc_server_test.go [hunk 8]
```diff
@@ -539,12 +539,12 @@ func TestGetLeavesByLeafValueHashInvalidHash(t *testing.T) {
...
 	}
 
 	// And failed at app level
 	if got, want := resp.Status.StatusCode, trillian.TrillianApiStatusCode_ERROR; got != want {
-		t.Fatalf("GetLeavesByLeafValueHash(hash empty) status = %d, want: %d", got, want)
+		t.Fatalf("GetLeavesByLeafValueHash(hash empty) = %v, %v, want: %d, %d", err, resp.Status, got, want)
```
Comment (daviddrysdale) [inline server/log_rpc_server_test.go]: Ditto

Comment (Martin2112) [inline server/log_rpc_server_test.go]: Both fixed.

**Comment by Martin2112** [comment]
Fixed.
